docs/_docs/20-fasterkv-basics.md

```diff
@@ -107,7 +107,95 @@ Apart from Key and Value, the IFunctions interface is defined on three additiona
+ReadCompletionCallback receives the `RecordMetadata` of the record that was read.
+
+##### Return Values of IFunctions Data Operations
+IFunctions calls receive data through reference parameters to Keys and their associated Values. As well, they receive a parameter of either `ReadInfo` or `UpdateInfo`, describing additional information about the operation and its context:
+- **ReadInfo**: Contains the session Version, session type, and logical address
+- **UpdateInfo**: Contains the session Version, session type, and logical address, as well as two aspects of the value length that are important for reuse of variable-length records, either by shrinking and re-growing the value, or by revivifying a tombstoned record:
+  - **FullValueLength**: The full size of the allocated value space.
+  - **UsedValueLength**: The amount of the allocated value space that is in use. This is initially set by FASTER to the FullValueLength, and is reset by the application's IFunctions callback as appopriate.
+  - **DeleteRecord**: Always False coming in from FASTER; may be set True by the app if, for example, expiration has been detected.
+  - **CancelOperation**: Always False coming in from FASTER; may be set True by the app if it detects a condition in the callback that makes the operation no longer desired.
+
+Additionally, the callback return values control some aspects of the operation, particularly related to:
+- The ***Required Length*** to insert or update a value exceeds the available value length
+- Record ***Expiration*** is detected
+- The application wants to ***Cancel*** the current operation
+
+**Read**:
+- `SingleReader`:
+  - ***Required Length***: Irrelevant because this is not an update
+  - ***Expiration***: The app sets readInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by forcing a CopyToTail of the current record with the Tombstone set, and Read will return (!status.Found && status.CreatedRecord).
+  - ***Cancellation***: The app sets readInfo.CancelOperation True and returns false. Read() returns (!status.Found).
+- `ConcurrentReader`
+  - ***Required Length***: Irrelevant because this is not an update
+  - ***Expiration***: The app sets readInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by in-place update, and Read will return (!status.Found).
+  - ***Cancellation***: The app sets readInfo.CancelOperation True and returns false. Read() returns (!status.Found).
+
+**RMW**
+  - `InPlaceUpdater`
+    - ***Required Length***: The app sets updateInfo.FullValueLength and updateInfo.UsedValueLength to the existing lengths and returns false. FASTER will allocate a record with the required length and call `CopyUpdater`.
+    - ***Expiration***: The app sets updateInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by in-place update, and RMW will return (!status.Found).
```

Comment:
in case item is set to deleted, why would the app return notfound? wouldnt we have to retry and go through the initial-updater code path? the operation is not canceled in this case.

Reply:
Doc updated: this does go to InitialUpdater unless Canceled *also*; this combination means "delete it and return", so RMW returns (status.NotFound | status.Record.InPlaceUpdated).